samanthas_telegram_bot/api_queries/smalltalk.py

```diff
@@ -0,0 +1,208 @@
+                )
+
+            logger.info(f"Chat {user_data.chat_id}: SmallTalk results not ready. Waiting...")
+            attempts += 1
+            await asyncio.sleep(SMALLTALK_TIMEOUT_IN_SECS_BETWEEN_ATTEMPTS)
+        else:
+            await log_and_notify(
+                bot=context.bot,
+                logger=logger,
+                level=LoggingLevel.INFO,
+                text=(
+                    f"Chat {user_data.chat_id}: Received [SmallTalk results for "
+                    f"{user_data.first_name} {user_data.last_name}]({result.url})"
+                ),
+                parse_mode_for_admin_group_message=ParseMode.MARKDOWN_V2,
+            )
+            return result
+
+
+async def get_json_with_results(test_id: str) -> bytes:
+    async with httpx.AsyncClient() as client:
+        response = await client.get(
+            url=SMALLTALK_URL_GET_RESULTS,
+            headers=HEADERS,
+            params={
+                "id": test_id,
+                "additional_fields": (
+                    "detailed_scores,strength_weaknesses,problem_statuses,problem_titles"
+                ),
+            },
+        )
+
+    logger.info(
+        f"Request headers: {response.request.headers}. "
+        f"Response: {response.status_code=}, {response.headers=}, {response.content=}"
+    )
+    return response.content
+
+
+def process_smalltalk_json(json_data: bytes) -> SmalltalkResult | None:
+    def level_is_undefined(str_: str) -> bool:
+        return str_.lower().strip() == "undefined"
+
+    try:
+        loaded_data = json.loads(json_data)
+    except json.decoder.JSONDecodeError:
+        logger.error(f"Could not load JSON from {json_data=}")
+        return None
+
+    status = loaded_data["status"]
+
+    if status not in SmalltalkTestStatus._value2member_map_:  # noqa
+        raise ApiRequestError(f"SmallTalk returned {status=} but we have no logic for it.")
+
+    if status == SmalltalkTestStatus.NOT_STARTED_OR_IN_PROGRESS:
+        logger.info("User has not yet completed the interview")
+    elif status == SmalltalkTestStatus.RESULTS_NOT_READY:
+        logger.info("User has completed the interview but the results are not ready")
+
+    if status != SmalltalkTestStatus.RESULTS_READY:
+        return SmalltalkResult(status=status)
+
+    level = loaded_data["score"]
+    level_id = level[:2]  # strip off "p" in "B2p" and the like
+
+    if level_is_undefined(level):
+        logger.info("User did not pass enough oral tasks for level to be determined")
+        level_id = None
+    elif level_id not in ALL_LEVELS:
+        logger.error(f"Unrecognized language level returned by SmallTalk: {level}")
+        level_id = None
+
+    results_url = loaded_data["report_url"]
+
+    logger.info(f"SmallTalk results: {status=}, {level=}, {results_url=}")
+
+    return SmalltalkResult(
+        status=status,
+        level=level_id,
+        url=results_url,
+        original_json=json_data,
```

Comment:
You already have `loaded_data` here. Why you keep bytes if you then again should do `json.dumps(str(...))`?

Reply:
Fixed